lib/bookmark_query.rb

```diff
@@ -54,27 +64,97 @@ def list_all
 
     results = results.limit(@limit)
 
+    BookmarkQuery.preload(results, self)
+    results
+  end
+
+  private
+
+  def list_all_results(topics, pms)
+    results = base_bookmarks.merge(topics.or(pms))
+    results = results.merge(Post.secured(@guardian))
+    results = @guardian.filter_allowed_categories(results)
+
     if BookmarkQuery.preloaded_custom_fields.any?
-      Topic.preload_custom_fields(
-        results.map(&:topic), BookmarkQuery.preloaded_custom_fields
-      )
+      Topic.preload_custom_fields(results, BookmarkQuery.preloaded_custom_fields)
     end
 
-    BookmarkQuery.preload(results, self)
-
-    @guardian.filter_allowed_categories(results)
+    results
   end
 
-  private
+  def list_all_results_polymorphic(topics, pms)
+    topic_results = topic_bookmarks.merge(topics.or(pms))
+
+    post_results = post_bookmarks.merge(topics.or(pms))
+    post_results = post_results.merge(Post.secured(@guardian))
+
+    topic_results = @guardian.filter_allowed_categories(topic_results)
+    post_results = @guardian.filter_allowed_categories(post_results)
```

Comment:
Actually I think this can be combined in to a single query and there is no need for a `UNION` . In `base_bookmarks_polymorphic`, all bookmarks of type `Post` and `Topic` will eventually be joined against the topics table which is then used to filtered away topics that the user does not have access to.

Reply:
They do actually all get combined into the single query with UNIONs, this main one:

```
SELECT bookmarks.* FROM (SELECT "bookmarks"."id", "bookmarks"."user_id", "bookmarks"."post_id", "bookmarks"."name", "bookmarks"."reminder_at", "bookmarks"."created_at", "bookmarks"."updated_at", "bookmarks"."reminder_last_sent_at", "bookmarks"."reminder_set_at", "bookmarks"."auto_delete_preference", "bookmarks"."pinned", "bookmarks"."for_topic", "bookmarks"."bookmarkable_id", "bookmarks"."bookmarkable_type" FROM "bookmarks" LEFT JOIN posts ON posts.id = bookmarks.bookmarkable_id AND bookmarks.bookmarkable_type = 'Post' LEFT JOIN topics ON topics.id = posts.topic_id OR (topics.id = bookmarks.bookmarkable_id AND bookmarks.bookmarkable_type = 'Topic') LEFT JOIN topic_users ON topic_users.topic_id = topics.id WHERE "bookmarks"."user_id" = 1 AND (topic_users.user_id = 1) AND "bookmarks"."bookmarkable_type" = 'Topic' AND "topics"."deleted_at" IS NULL AND ((topics.archetype <> 'private_message') AND (topics.category_id IS NULL OR topics.category_id IN (SELECT id FROM categories WHERE NOT read_restricted OR id IN (3,4,5,9,16,17,20,23,29,34))) OR "topics"."archetype" = 'private_message' AND (topics.id IN (SELECT topic_id
FROM topic_allowed_users
WHERE user_id = 1
)
      OR topics.id IN (SELECT tg.topic_id
FROM topic_allowed_groups tg
JOIN group_users gu ON gu.user_id = 1 AND gu.group_id = tg.group_id
))) UNION SELECT "bookmarks"."id", "bookmarks"."user_id", "bookmarks"."post_id", "bookmarks"."name", "bookmarks"."reminder_at", "bookmarks"."created_at", "bookmarks"."updated_at", "bookmarks"."reminder_last_sent_at", "bookmarks"."reminder_set_at", "bookmarks"."auto_delete_preference", "bookmarks"."pinned", "bookmarks"."for_topic", "bookmarks"."bookmarkable_id", "bookmarks"."bookmarkable_type" FROM "bookmarks" LEFT JOIN posts ON posts.id = bookmarks.bookmarkable_id AND bookmarks.bookmarkable_type = 'Post' LEFT JOIN topics ON topics.id = posts.topic_id OR (topics.id = bookmarks.bookmarkable_id AND bookmarks.bookmarkable_type = 'Topic') LEFT JOIN topic_users ON topic_users.topic_id = topics.id WHERE "bookmarks"."user_id" = 1 AND (topic_users.user_id = 1) AND "bookmarks"."bookmarkable_type" = 'Post' AND "topics"."deleted_at" IS NULL AND ((topics.archetype <> 'private_message') AND (topics.category_id IS NULL OR topics.category_id IN (SELECT id FROM categories WHERE NOT read_restricted OR id IN (3,4,5,9,16,17,20,23,29,34))) OR "topics"."archetype" = 'private_message' AND (topics.id IN (SELECT topic_id
FROM topic_allowed_users
WHERE user_id = 1
)
      OR topics.id IN (SELECT tg.topic_id
FROM topic_allowed_groups tg
JOIN group_users gu ON gu.user_id = 1 AND gu.group_id = tg.group_id
))) AND "posts"."deleted_at" IS NULL AND (posts.post_type IN (1,2,3,4)) UNION SELECT "bookmarks"."id", "bookmarks"."user_id", "bookmarks"."post_id", "bookmarks"."name", "bookmarks"."reminder_at", "bookmarks"."created_at", "bookmarks"."updated_at", "bookmarks"."reminder_last_sent_at", "bookmarks"."reminder_set_at", "bookmarks"."auto_delete_preference", "bookmarks"."pinned", "bookmarks"."for_topic", "bookmarks"."bookmarkable_id", "bookmarks"."bookmarkable_type" FROM "bookmarks" WHERE "bookmarks"."user_id" = 1 AND "bookmarks"."bookmarkable_type" NOT IN ('Post', 'Topic')) as bookmarks ORDER BY (CASE WHEN bookmarks.pinned THEN 0 ELSE 1 END),
        bookmarks.reminder_at ASC,
        bookmarks.updated_at DESC LIMIT 20; 
```

Is it because `filter_allowed_categories` and `Post.secured` do different things, and the latter has to be run on the posts.